### ISSUE TITLE
[MNG-5761] Dependency management is not transitive.

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7914]
**[Jeff Schnitzer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lhoriman)** opened **[MNG-5761](https://issues.apache.org/jira/browse/MNG-5761?redirect=false)** and commented

A detailed description of the issue is here:

http://stackoverflow.com/questions/28312975/maven-dependencymanagement-version-ignored-in-transitive-dependencies

The short of it is that maven appears to be using the wrong \<dependencyManagement> version in a transitive dependency.  There are two relevant \<dependencyManagement> sections in the build, one pulled in by guice and one pulled in by gwizard-parent. These are the dependency paths from the top:

gwizard-example -> gwizard-config -> gwizard-parent
gwizard-example -> gwizard-config -> guice -> guice-parent

gwizard-parent's dependencyManagement specifies guava 18
guice-parent's dependencyManagement specifies guava 16

Guava 16 is winning. This seems highly undesirable, and in fact it breaks our build. I would expect that in a version # fight, "closest to the top" should win.

---

**Affects:** 3.2.5

**Attachments:**
- [depending-pom2.xml](https://issues.apache.org/jira/secure/attachment/13071971/depending-pom2.xml) (_737 bytes_)
- [MNG-5761.zip](https://issues.apache.org/jira/secure/attachment/12774550/MNG-5761.zip) (_1.70 kB_)

**Issue Links:**
- [MRESOLVER-10](https://issues.apache.org/jira/browse/MRESOLVER-10) New 'TransitiveDependencyManager' supporting transitive dependency management
 (_**"is blocked by"**_)
- [MNG-6056](https://issues.apache.org/jira/browse/MNG-6056) Implement Feature Toggle Module to handle Feature Toggles
 (_**"is blocked by"**_)
- [MNG-3890](https://issues.apache.org/jira/browse/MNG-3890) Transitive dependencies override explicitly set scope.
 (_**"duplicates"**_)
- [MNG-5769](https://issues.apache.org/jira/browse/MNG-5769) Maven is ignoring dependencyManagement for transitive dependencies
 (_**"is duplicated by"**_)
- [FLINK-9091](https://issues.apache.org/jira/browse/FLINK-9091) Failure while enforcing releasability in building flink-json module
 (_**"causes"**_)
- [MNG-8295](https://issues.apache.org/jira/browse/MNG-8295) Dependency Manager Transitivity (now default) handles dependency management inconsistently

- [MNG-7003](https://issues.apache.org/jira/browse/MNG-7003) Inconsistent dependency tree

- [MNG-4720](https://issues.apache.org/jira/browse/MNG-4720) Create an integration test to capture effects of exclusions from dependency management in combination with exclusions from regular dependencies

- [MNG-7982](https://issues.apache.org/jira/browse/MNG-7982) Switch to enable transitivity in depMgr used by Maven

- [MNG-6139](https://issues.apache.org/jira/browse/MNG-6139) Addition of command line option 'legacy-dependency-management'.
 (_**"is required by"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=67636510)

**Backported to:** [4.0.0-beta-5](https://github.com/apache/maven/milestone/24?closed=1)

14 votes, 39 watchers

**Comment by jira-importer** [comment]
**[Maria Petridean](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maria.petridean)** commented

The description of this issue here, does not match 100% the description on stackoverflow. I encountered the same issue: a project A with a parent pom which has a dependencyManagement section (everything ok until now). Then, another project B, depends on project A and the result is that the dependencyManagement section in A'a parent pom is ignored for resolving the transitive dependencies.

Could you update the description of this issue to reflect also this case?

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Please provide a sample problem which allows us the reproduce the issue.

**Comment by jira-importer** [comment]
**[Maria Petridean](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maria.petridean)** commented

[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl) this is reproducible.

For example:

```xml
<dependencyManagement>
        <dependencies>
            <dependency>
				<groupId>...</groupId>
				<artifactId>project_T</artifactId>
				<version>1.0</version>
			</dependency>
		</dependencies>
</dependencyManagement>
```

```xml
...
<parent>
    <groupId>...</groupId>
    <artifactId>parent_pom</artifactId>
    <version>...</version>
</parent>
....
<dependencies>
	<dependency>
		<groupId>...</groupId>
		<artifactId>project_B</artifactId>
		<version>...</version>
	</dependency>
</dependencies>
```

```xml
<dependencies>
	<dependency>
		<groupId>...</groupId>
		<artifactId>project_T</artifactId>
		<version>2.0</version>
	</dependency>
</dependencies>
```

```xml
<dependencies>
	<dependency>
		<groupId>...</groupId>
		<artifactId>project_A</artifactId>
		<version>...</version>
	</dependency>
</dependencies>
```

Now:
1. running `mvn dependency:tree` from project_A will result in bringing version 1.0 of project_T, which is correct.
2. running `mvn dependency:tree` from project_C will result in bringing version 2.0 of project_T, which is incorrect (the same dependency management rules should apply as at point 1, which is the expected.).

This same scenario, is also explained here http://stackoverflow.com/questions/28312975/maven-dependencymanagement-version-ignored-in-transitive-dependencies.

**Comment by jira-importer** [comment]
**[Jeff Schnitzer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lhoriman)** commented

I can appreciate the desire to be spoon-fed a minimal test case; we're all busy people. However, isn't closing this issue rather premature?

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

DefaultDependencyManager goes back to Maven 3.4, which never got released. So this is what should have been the default behaviour from day one. I am no longer doing Java development. To make a long story short. I took a look at the Jira issues somewhen 2015 or so - one of those you seem to have reopened - and just fixed inconsistencies, bugs and such. This was not possible due to integration tests starting to fail - no matter if the test was testing correct behaviour or not. It was then decided to reset almost all of my work and retain behaviour - even if inconsistent or just not behaving as advertised. I kept a 3.4-SNAPSHOT version in my private Github repository. You can download it from there and you will soon get, why nothing of that will ever land in an official release. When it was clear, there will be no more changes in resolution behaviour for pom 4.0, I added some warning messages as well. That's the version I did run locally somewhen 7 years ago and I would run locally, when I would still be doing Java things. I do not even know, if it still works at all. I read the notification mail by accident, when configuration my spam filter.

[3.4-SNAPSHOT](https://github.com/ChristianSchulte/maven/releases/tag/2017-11-27)

**Comment by jira-importer** [comment]
**[Didier Loiseau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=didier%20loiseau)** commented

`@Everyone`, now that the fix for MNG-8295 has been merge, you’ll be glad to read that, with the current master, you’ll get the expected output:

```java
$ mvn -V -f depending-pom.xml dependency:tree
Apache Maven 4.0.0-beta-5-SNAPSHOT (a293822852c90ca7cca636d11ed0d7a6c7eda691)
[…]
[INFO] --- dependency:3.8.0:tree (default-cli) @ depending ---
[INFO] MNG-5761:depending:pom:1.0-SNAPSHOT
[INFO] \- MNG-5761:dependent:pom:1.0-SNAPSHOT:compile
[INFO]    \- commons-beanutils:commons-beanutils:jar:1.9.2:compile
[INFO]       +- commons-logging:commons-logging:jar:1.1.1:compile
[INFO]       \- commons-collections:commons-collections:jar:3.2.2:compile
```

[gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet), [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas), I think this one can be marked as resolved in 4.0.0-beta-5

[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77) While I understand your bitterness, I think you are wrong. Things seem to be changing. `<modelVersion>4.1.0</modelVersion>` _is_ a thing {_}now{_}. It is true that there is a huge passive with all artifacts published in Maven Central and other repositories, but they have found a solution for this as well: consumer poms.

I have been using Maven for 14+ years, but it is indeed the first time I get involved in its development. In fact, I’m pretty sure I’ve come across this dependency management issue in the past. And while I’m a bit worried that it could still somehow get reverted, I’m hopeful. In the worst case, I would have lost a few hours of my time, and learned a bit about Maven internals, which can’t be bad either.

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte)** commented

[Didier Loiseau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Didier%20Loiseau) Indifference, not bitterness.

parent
- module1
- module2

Build in parent, get a different dependency tree. Build in module1, get a different dependency tree. Build in module2, get a different dependency tree. What build system is that, not fixing a key feature for more than a decade and not even now. Requested by various users independently for Maven 2, Maven 3, Maven 4 and what not. Repeatedly reported. Repeatedly closed won't fix. Repeatedly reopened. Ridiculous. Has driven away lots of users. Those said "millions of users" never existed. Received a lot of justified rants for things like this.
